### DISCUSSION TITLE
Joint vision: Future GliderTools package structure

### DISCUSSION THREAD

**Discussion by soerenthomsen** [GliderToolsCommunity/GliderTools#107] [General]
Hi everyone, 

As promised I worked on a draft figure for a new package structure. Some of you have already contributed to this proposal. Thanks a lot for your input. I tried to follow the data lifecycle (left to right). You can click on the figure to expand it. 

<img width="853" alt="Screen Shot 2021-04-07 at 15 47 52" src="https://user-images.githubusercontent.com/32633545/113877085-a56b9580-97b8-11eb-8f4e-80af8e7cbe2c.png">

At this stage the main goal is to develop a joint vision. 
(1) So please feel free to comment on it
(2) State where you might want to contribute. 
(3) Do you already have a software package which could be implemented or made compatible with GliderTools?
(4) Are you working on something right now which could be aligned with this community effort? Maybe you can get support?

If you want to make direct changes to the figure (within the MIRO board), please let me know. I can give you access. 

Best Soeren
@lukegre @jbusecke @dhruvbalwada @callumrollo @isgiddy @tjryankeogh @marcelduplessis @tomhull @bastienqueste @AnthonyBosse @castelao @patricialg @sarahnicholson @sebswart @AlicePietri @dennisstanley64 @sryan288 @LewisDrysdale @backeb @nlbeaird

**Comment by callumrollo** [comment]
Related tool: I'll be working on [gliderpy](https://github.com/ioos/gliderpy) this summer. I'd love to make it easy to pass glider data downloaded from an ERDDAP server into GliderTools

**Reply by dhruvbalwada** [reply]
That would be awesome!!!

**Comment by jbusecke** [comment]
From the discussion today it seems to me (might be overlooking something here), that we should clarify (and perhaps modify) they representation of gliderdata that glidertools expects. 

It seems that the most 'raw'/general way to pass data is in form of a 'timeseries' or 1D array (where time is the only real dimension, and other values like depth/lon/lat ... are coordinates). A common 'higher level' product is a profile 2D array that has some form of depth coordinate and a profile index. Then there are even higher level 'derived' products (mapped 3/4D data), but these are not easily convertible back and forth, so I will exclude them here.

From @callumrollo excellent demo notebook I take that glidertools commonly uses the former 1D representation, but certain applications might favor the 2D one (@castelao is this echoing some of your concerns you raised during the call today?).

glidertools is nicely positioned to do this conversion, and assure that the roundtrip conversion does not loose any data. Tools that build on that could then choose between either representation of the data. 

@kerfoot mentioned that his tool was able to convert between these two? Perhaps we could refactor that into the main glidertools package (and move some of the I/O related code from there to either your package or another intermediate one.

**Reply by dhruvbalwada** [reply]
We have an open issue that might be related: https://github.com/GliderToolsCommunity/GliderTools/issues/30

**Reply by castelao** [reply]
@jbusecke , you mentioned a plugin-based architecture during the meeting, which I like very much. I suggest even further, not quite plugins but an ecosystem where the relatively independent packages can communicate with each other. It's almost the same thing since each plugin, or package, requires its own documentation, consistency tests, etc, but in a more horizontal structure. With a clear data model, the components could communicate with each other. A parser package could read some raw dataset and feed directly the plotting package, while in another scenario could be convenient to have the real-time QC in the middle, or ...

> applications might favor the 2D one

Yes. For some procedures, it is more convenient to operate in profiles, or a full dive, which can be easily obtained (reshaping + groupby) from the 1D object, if there is a standard structure.

I suggest having the I/O as separate packages that bring each platform as close as possible to a common structure. Different objectives can require different information to be extracted from the raw formats, and a full extraction can get a little complex while being exclusive for that specific raw format. For Spray, I developed an internal tool (SprayTools) that reads our raw format or SQL or the internal MatLab format or ... and returns a standard xarray object. My suggestion is to isolate the possible complexities on where they might be required, especially if they are exclusive.

**Reply by jbusecke** [reply]
I think this is a great way forward @castelao. I have outlined some possible steps in #129, to move this forward. I think having some actual examples of e.g. your workflow and how glidertools could help to use your work would be amazing.

**Comment by soerenthomsen** [comment]
**Meeting notes from GliderTools / OceanGliders Best Practice Workshop**
May 24 Monday - 16:00 - 17:30 CEST / 7:00 - 8:30 PDT / 22:00 - 23:30 AWST 

Session chairs: I. Giddy, T. Ryan-Keogh, D. Balwada, C. Rollo, S. Thomsen, S. Nicholson

Discuss on future vision of package structure and overall GliderTools community.

**Participants** 
Soeren Thomsen (https://github.com/ThomsenSoeren/) 
Dan Hayes, 
Gui Castelão (https://github.com/castelao/)
Sarah Nicholson (https://github.com/sarahnicholson)
Thomas Ryan-Keogh (https://github.com/tjryankeogh/) 
Callum Rollo (https://github.com/callumrollo/)
Carolina Amadio, 
John Kerfoot (https://github.com/kerfoot/) 
Isabelle Giddy (https://github.com/isgiddy/) 
María Valladares, 
Yumi Song, 
Dhruv Balwada (https://github.com/dhruvbalwada)
Tom Hull (https://github.com/tomhull/) 
Filipa Carvalho, 
Brita Irving, 
Julius Busecke (https://github.com/jbusecke/) 
Nikolaos Zarokanellos (https://github.com/nizaroka/)
Alice Pietri (https://github.com/AlicePietri/) 

Link to google doc
https://docs.google.com/document/d/1ZlcO5984hgzkts9vwvXvEwHve6c2kZUwbvzWGv-TmCI/edit?pli=1

**Reply by soerenthomsen** [reply]
@bastienqueste, @marcelduplessis, @lukegre, @smerckel, @AnthonyBosse
you can read the meeting notes of our discussion on the future vision of the GliderTools package and GliderTools community in this thread.

**Comment by dhruvbalwada** [comment]
One of the question that came in was better defining the focus of Glidertools. 
This is something we should state clearly at: https://glidertools.readthedocs.io/en/latest/ 
Right now it reads that "Glider tools is a Python 3.6+ package designed to process data from the first level of processing to a science ready dataset." . This statement as it stands is a bit vague (what does first level of processing mean? Do we mean what base station spits out? What does science-ready dataset mean? etc). Also it does not cover one of the very important points that Glidertools allows you (and will allow more as options get added) to do a lot of the science (visualization being part of it).

If people post here what they think the focus of GT is and should be, in broad terms, then I can help rewrite the intro page. 

For example someone raised the point that they want to allow people who have no coding or oceanography experience to be able to explore the data. I think that is a great use case, but definitely not something that we are trying or want to achieve with GT. In my opinion:
Our focus is more for people looking to do custom analysis, who are familiar with some basic coding and are willing to write code. It allows these users to use GT to do some basic things (DMQC, visualize, grid, secondary variables, etc) quickly, and then build on the analysis using all of the other python ecosystem. We are not looking to create a standalone platform/toolbox that does some common select operations and then having to move the data over to other analysis languages. Also, in the future if someone has the heart and energy to add some operations to GT, then they can code that, and it can be easily merged after review.

**Comment by dhruvbalwada** [comment]
A second vision for the Glidertools community (not GT package) can be documenting the process for doing different analyses. Callum's demo notebooks show how this can be done for some of the options available in GT (https://github.com/GliderToolsCommunity/glidertools-demo). However, we can even think bigger and create a place for people to share notebooks that use GT as the base (or not), but then have to use other packages to achieve other specific goals (e.g. I can think of @isgiddy spectral analysis as one example). In the long term, this can be very helpful to new users of gliders.

**Reply by soerenthomsen** [reply]
I like this idea too. It helps other not so visible efforts to get more attention too.

**Reply by jbusecke** [reply]
This could play well together with [pangeo gallery](https://gallery.pangeo.io), where we could host a variety of 'real world' examples. I recommend keeping those in a separate repo. 
I have done something like that with [xgcm](https://github.com/xgcm/xgcm) and [xgcm-examples](https://github.com/xgcm/xgcm-examples). 
Over there these are also rendered out into our documentation (but that is totally optional).
